lib/engine/src/export.rs

```diff
@@ -44,17 +46,57 @@ impl From<VMExport> for Export {
     }
 }
 
+/// Extra metadata about `ExportFunction`s.
+///
+/// The metadata acts as a kind of manual virtual dispatch. We store the
+/// user-supplied `WasmerEnv` as a void pointer and have methods on it
+/// that have been adapted to accept a void pointer.
+#[derive(Debug, PartialEq)]
+pub struct ExportFunctionMetadata {
+    /// This field is stored here to be accessible by `Drop`.
+    ///
+    /// At the time it was added, it's not accessed anywhere outside of
+    /// the `Drop` implementation. This field is the "master copy" of the env,
+    /// that is, the original env passed in by the user. Every time we create
+    /// an `Instance` we clone this with the `host_env_clone_fn` field.
+    ///
+    /// Thus, we only bother to store the master copy at all here so that
+    /// we can free it.
+    ///
+    /// See `wasmer_vm::export::VMExportFunction::vmctx` for the version of
+    /// this pointer that is used by the VM when creating an `Instance`.
+    pub host_env: *mut std::ffi::c_void,
```

Comment:
Idea: Create a `new` constructor method rather than having all fields `pub`. I think it could be “dangerous”, it's better to hide everything here. Thoughts?

Reply:
Same for all the fields in this type.